Extend support for more kinds of barriers, including node and ways, and blocking pedestrians / bicycles in addition to motor vehicles

miklcct opened the pull request:
### Summary

Improve support for more kinds of barriers, including point barriers and linear barriers, which can block travel for multiple modes of traversal.

In implementation terms, it does the following:

- Barrier node detection is improved to handle more kinds of barriers and traversal modes
- If a way crosses a linear barrier, the vertices are split with an edge added to cross the barrier only for allowable traversal modes. If the node is tagged with `barrier=entrance`, the restrictions for the barrier is ignored. <s>Note that if two linear barriers join end to end, it is possible to cross between them at the joining node, unless the node itself blocks access.</s> Update 2025-08-06: After checking some real-world mapping, I have decided to disallow passing through the end of a barrier as well, like the photo shown below but with two areas.
- Areas are not joined if they share the same barrier which blocks traversal for some street modes for the area (sharing at least two nodes of the barrier), so that a linear barrier can block area routing between two areas and force the user to go via the end. Also, areas are only joined if they share at least two consecutive nodes (when two areas join at one node, it is possible for a wall to block access between them).

#### What this PR does not do ####
Note that barrier blocking access within a single area (whether the barrier effectively splits the area into two or not), like the following picture is not implemented.
<img src="https://wiki.openstreetmap.org/w/images/1/14/Barrier_and_waypoint-EN.png" alt="barrier in area"/>

### Issue

Closes #6689

The following screenshots show the fixed routing to avoid walking into a wall inside a train station:

original:
<img width="1920" height="1080" alt="original" src="https://github.com/user-attachments/assets/e7640c84-2e29-4ee6-ab08-bd43b85d8636" />
fixed:
<img width="1920" height="1036" alt="image" src="https://github.com/user-attachments/assets/f071a676-355d-4165-9382-3d78900a1b10" />


### Unit tests

Added for area splitting, vertex handling and barrier permission detection including wheelchair accessibility.

### Documentation

### Changelog

### Bumping the serialization version id

Definitely required. A new kind of vertex is added to prevent duplicate labels for split vertices.

miklcct commented:
I have just sent an email to OSM tagging mailing list to ask for behavior for routing through the end of a linear barrier.

miklcct reviewed (commented):
Add comment on barrier after OSM ground survey

miklcct commented on application/src/main/java/org/opentripplanner/graph_builder/module/osm/OsmAreaGroup.java:
```diff
@@ -86,29 +88,46 @@ public OsmAreaGroup(Collection<OsmArea> areas) {
-          if ((level1 == null && level2 == null) || (level1 != null && level1.equals(level2))) {
+          boolean onSameLevel =
+            (level1 == null && level2 == null) || (level1 != null && level1.equals(level2));
+          boolean shareBarrier =
+            area1 != area2 &&
+            !CollectionUtils.intersection(
+              barriersForArea.get(area1),
+              barriersForArea.get(area2)
+            ).isEmpty();
```
Yes. It will prevent merging as you don't want cars to drive through the bollards.

miklcct commented on application/src/main/java/org/opentripplanner/graph_builder/module/osm/VertexGenerator.java:
```diff
@@ -121,6 +133,19 @@ IntersectionVertex getVertexForOsmNode(OsmNode node, OsmEntity way) {
     return iv;
   }
 
+  private IntersectionVertex getSplitVertexOnBarrier(OsmNode node, OsmEntity way) {
```
A vertex on a barrier is split into vertices, one for each way it belongs to.

miklcct replied:
I have added a Javadoc to explain it

miklcct commented on application/src/main/java/org/opentripplanner/graph_builder/module/osm/OsmAreaGroup.java:
```diff
@@ -86,29 +88,46 @@ public OsmAreaGroup(Collection<OsmArea> areas) {
-        }
-        for (OsmNode node : ring.nodes) {
-          areasForNode.put(node, area);
+          processRing(area, inner, barriers, areasForNodePair, barriersForArea);
```
Yes. Imagine two areas, one inside the other, and the inside area is partially surrounded by barriers. We don't want them to join together.

miklcct commented on application/src/main/java/org/opentripplanner/street/model/vertex/StationEntranceVertex.java:
```diff
@@ -50,7 +50,7 @@ public Accessibility wheelchairAccessibility() {
   public String toString() {
     return ToStringBuilder.of(StationEntranceVertex.class)
-      .addNum("nodeId", nodeId)
+      .addNum("wayId", nodeId)
```
It was a mistake

miklcct commented on application/src/main/java/org/opentripplanner/street/model/vertex/OsmVertexOnWay.java:
```diff
@@ -0,0 +1,18 @@
+/**
+ * This vertex is created for a node on a particular way
+ */
+public class OsmVertexOnWay extends OsmVertex {
```
Do you have any suggestions?

vesameskanen replied:
BarrierPassThroughVertex?

vesameskanen replied:
Well, those vertices are generated to enable traversal trough a barrier. They are not on a side of a barrier but exactly on the barrier. I think my proposal is more descriptive than yours.

vesameskanen reviewed (changes requested):
Regression is indeed solved, good work! However, I have some further concerns:

- I examined what kind of new connections appear through barriers. Almost all seem to be added to barrier=kerb holes inside areas. Those kerb holes are not routable (area processing works that way), so added edges serve no purpose at all. An example: https://www.openstreetmap.org/way/223417527 . Would it be possible to exclude such edges? I believe an edge should lead to another routable area in order to make sense.
- It seems that in most cases, area merging is prevented by barriers outside the areas, for example: Barrier blocks merging areas https://www.openstreetmap.org/way/624169127 and https://www.openstreetmap.org/way/28654317 . I guess there is a need to make some sort of geometrical test to fix such errors.

miklcct commented:
In both of your cases, there are multiple areas overlapping each other. 

I have no idea they actually exist in the real world and I have no idea what the behaviour should be, especially if the areas partially overlap each other and there is a barrier crossing the overlapped area.

vesameskanen commented:
Overlapping is not the point, although it surely may complicate things.
More examples: 

https://www.openstreetmap.org/way/488289620 and https://www.openstreetmap.org/way/30368517
https://www.openstreetmap.org/way/1136165707 and https://www.openstreetmap.org/way/1136165699

In the last example, there is a barrier which runs on the side of the two areas:
https://www.openstreetmap.org/way/692477220#map=19/60.182488/24.981725&layers=N

Apparently the code should examine if barriers actually block something.

miklcct commented:
> Overlapping is not the point, although it surely may complicate things. More examples:
> 
> https://www.openstreetmap.org/way/488289620 and https://www.openstreetmap.org/way/30368517 https://www.openstreetmap.org/way/1136165707 and https://www.openstreetmap.org/way/1136165699
> 
> In the last example, there is a barrier which runs on the side of the two areas: https://www.openstreetmap.org/way/692477220#map=19/60.182488/24.981725&layers=N
> 
> Apparently the code should examine if barriers actually block something. Hopefully this does not introduce a high time penalty, as geometric intersection tests tend to be slow.

Thanks, that's not intended. My intention is to only block merging if two areas share the same two nodes on a barrier.

vesameskanen commented:
Could it be so that barrier should have 2 consecutive nodes shared by the areas to prevent merging ? Now it seems that a single shared barrier node is enough.

vesameskanen reviewed (commented):
I am seeing good progress here. Barrier processing no longer increases graph build time, quite the opposite: it is now faster than before. Maybe because you removed some double edges on area boundaries.

Anyway, I can still see some changes which need further consideration:

1. Two areas are no longer connected here, fore good reason: https://www.openstreetmap.org/node/4945615563 . However, the second area drops from the graph which appears wrong, as gates allow access to it. See the images below.
<img width="704" height="704" alt="before" src="https://github.com/user-attachments/assets/f96f382c-c714-4f02-9895-709ce4f7d37c" />
<img width="665" height="656" alt="after" src="https://github.com/user-attachments/assets/b63c877d-4672-4772-b3a6-54ab38ab2f15" />
Maybe barrier=gate should be be considered just like barrier=entrance? 

2. I think we need to carefully analyze the requirement of adjacent node pair on barrier to activate blocking in area merge. It may be that there is 10 times more valid mergeable areas which share only discrete nodes than crappy barrier modeling errors. We should choose the logic which creates least errors.

vesameskanen commented on application/src/main/java/org/opentripplanner/street/model/vertex/VertexLabel.java:
```diff
@@ -59,6 +59,20 @@ public String toString() {
+   * be written sensibly.
+   */
+  record NodeOnWayLabel(long nodeId, long wayId) implements VertexLabel {
+    private static final String TEMPLATE = "osm:node:%s:way:%s";
+
```
I accidentally noticed that 'way' can be a relation as well. Is there a bug in logic, or is this just a minor naming issue?

miklcct replied:
This is just a naming issue, I think it happens elsewhere as well.

vesameskanen reviewed (changes requested):
I tested area merging in Helsinki region using modified  area merge blocking rules. The modified code is here:

https://github.com/HSLdevcom/OpenTripPlanner/tree/pairblock 

Here is the difference log:

* barrier blocks merge of https://www.openstreetmap.org/way/1134308217 and https://www.openstreetmap.org/way/1134308216
* barrier blocks merge of https://www.openstreetmap.org/way/1136165707 and https://www.openstreetmap.org/way/1136165708
barrier blocks merge of https://www.openstreetmap.org/relation/3150370 and https://www.openstreetmap.org/relation/14240934
barrier blocks merge of https://www.openstreetmap.org/relation/8661208 and https://www.openstreetmap.org/way/1093935304
barrier blocks merge of https://www.openstreetmap.org/way/1094525099 and https://www.openstreetmap.org/way/1094525103
barrier blocks merge of https://www.openstreetmap.org/way/1094530712 and https://www.openstreetmap.org/way/36813157
barrier blocks merge of https://www.openstreetmap.org/way/1134275948 and https://www.openstreetmap.org/way/1067689906
barrier blocks merge of https://www.openstreetmap.org/way/1134275948 and https://www.openstreetmap.org/way/1134275953
barrier blocks merge of https://www.openstreetmap.org/way/1183732929 and https://www.openstreetmap.org/way/1183732930
barrier blocks merge of https://www.openstreetmap.org/way/129204290 and https://www.openstreetmap.org/way/129204286
barrier blocks merge of https://www.openstreetmap.org/way/1314910974 and https://www.openstreetmap.org/way/1314910973
barrier blocks merge of https://www.openstreetmap.org/way/1315358556 and https://www.openstreetmap.org/relation/18042041
barrier blocks merge of https://www.openstreetmap.org/way/1315368869 and https://www.openstreetmap.org/way/1134275953
barrier blocks merge of https://www.openstreetmap.org/way/36813157 and https://www.openstreetmap.org/way/1094530712
barrier blocks merge of https://www.openstreetmap.org/way/457493348 and https://www.openstreetmap.org/way/1092837651

First two a badly modeled barriers in a same place. All other changes refer to places, where area merging should happen I strongly suggest that barrier blocks merging only when two adjacent nodes in both areas belong to it.

miklcct commented on application/src/main/java/org/opentripplanner/graph_builder/module/osm/BarrierEdgeBuilder.java:
```diff
@@ -27,7 +27,7 @@ public void build(OsmNode node, Collection<OsmVertex> vertices, Collection<OsmWa
     I18NString name = null;
 
-    if (!node.isTag("barrier", "entrance")) {
+    if (!node.hasTag("barrier")) {
```
What I have done is to consider all barrier=* on a node a hole in the barrier way, so the node barrier permission will apply. For example, if the node is a bollard, pedestrians and bikes and go through but cars can't.

vesameskanen commented:
Graph build seems  to work fine now. In Helsinki area, barriers block 182 area merges.

vesameskanen commented on application/src/main/java/org/opentripplanner/graph_builder/module/osm/OsmAreaGroup.java:
```diff
@@ -175,31 +196,27 @@ private static void processRing(
         }
+        var nodesOnBarrier = barrierMap.get(barrier);
+        nodesOnBarrier.add(new OsmNodePair(node, nextNode));
       }
```
This is really a minor issue, but can't we use already constructed node pair object `pair `here as well?

vesameskanen commented:
One possible improvement: we could add level information to OsmVertexOnWay and detect level conflicts when connecting the ways in BarrierEdgeBuilder. Conflicts could be reported as graph build issues, so that we get visibility to the magnitude of this problem.

vesameskanen commented on application/src/main/java/org/opentripplanner/graph_builder/module/osm/BarrierEdgeBuilder.java:
```diff
@@ -0,0 +1,75 @@
+
+    permission = node.overridePermissions(permission);
+    if (permission.allowsNothing()) {
+      return;
```
I agree.

vesameskanen commented on application/src/main/java/org/opentripplanner/graph_builder/module/osm/VertexGenerator.java:
```diff
@@ -95,24 +123,51 @@ IntersectionVertex getVertexForOsmNode(OsmNode node, OsmEntity way) {
+
+      if (iv == null && (node.isBarrier() || isNodeOnLinearBarrier)) {
+        iv = vertexFactory.barrier(nid, coordinate, node.explicitWheelchairAccessibility());
+      }
+
+      if (iv instanceof BarrierVertex bv) {
```
It seems that these 2 blocks can  be merged, right? iv is known to b e a barrierVertex here.

miklcct replied:
iv can be a BoardingLocation

miklcct replied:
I am also changing StationEntranceVertex to a subclass of BarrierVertex as well so they can't be merged.

vesameskanen commented on application/src/main/java/org/opentripplanner/graph_builder/module/osm/VertexGenerator.java:
```diff
@@ -55,10 +73,15 @@ public VertexGenerator(
    */
-  IntersectionVertex getVertexForOsmNode(OsmNode node, OsmEntity way) {
+  IntersectionVertex getVertexForOsmNode(
+    OsmNode node,
```
I wonder if getVertexForOsmNode should ber overloaded by 2 versions, one of which does not require linearBarrierNodeType. Now the type is visible in places where it is not relevant at all, like WalkableAreaBuilder.

miklcct replied:
WalkableAreaBuilder is the place where it is relevant - it is requesting to split nodes on linear barriers.

vesameskanen commented on application/src/main/java/org/opentripplanner/graph_builder/issues/DifferentLevelsSharingBarrier.java:
```diff
@@ -0,0 +1,21 @@
+
+public record DifferentLevelsSharingBarrier(OsmNode node) implements DataImportIssue {
+  private static final String FMT =
+    "Node %d is a barrier node but ways on different layers / levels are connected to it. Please check if the barrier actually blocks traversal, if not, please remove the node from the barrier.";
+
```
Please add an OSM link to the issue, so that it is easier to check the node:

```
@Override
  public String getHTMLMessage() {
    return String.format(HTMLFMT, node.url(), node.getId());
  }
```

vesameskanen reviewed (changes requested):
Issue reporting is a good improvement, thanks for it! The capital area of Finland alone generates 400 such issues. It may be that OSM validators do not care about layer conflicts and obviously we will never be able to fix the OSM. Right now, this PR generates 4 times more graph errors than it fixes.  Therefore I suggest that barrier edge builder should not block traversal in unclear conflict cases. Instead, OSM contributors can check the issues and start fixing badly modeled barriers.

vesameskanen reviewed (commented):
Graph build works now much better. Barriers no longer remove important bridges in Helsinki.

vesameskanen commented on application/src/main/java/org/opentripplanner/graph_builder/module/osm/VertexGenerator.java:
```diff
@@ -67,6 +80,10 @@ IntersectionVertex getVertexForOsmNode(OsmNode node, OsmEntity way) {
+    if (nodesInBarrierWays.containsKey(node)) {
+      return getSplitVertexOnBarrier(node, way);
+    }
     // single-level case
```
I added some logging and found out that BarrierEdgeBuilder tries to connect distant sections of area groups, not just the two at opposite side of a barrier.

I believe use of this entity reference is the reason:

https://github.com/opentripplanner/OpenTripPlanner/blob/ec60f14d3fe6850f7114e8bcfd0606027d3ad64e/application/src/main/java/org/opentripplanner/graph_builder/module/osm/WalkableAreaBuilder.java#L191

The code should consistently use parent entity of the currently processed area, not a random pick from the area group.

vesameskanen replied:
@miklcct I fixed this problem in https://github.com/opentripplanner/OpenTripPlanner/pull/6826

vesameskanen commented:
One more discovery: you should consider no through traversal limitations when adding connections through a barrier, the same way as here: 

https://github.com/HSLdevcom/OpenTripPlanner/blob/fix-area-edge-props/application/src/main/java/org/opentripplanner/routing/linking/VertexLinker.java#L744

It areas have private access and the connecting edges are not marked as no through, traversal is not possible - the edges are added for no reason.

vesameskanen commented on application/src/main/java/org/opentripplanner/graph_builder/module/osm/VertexGenerator.java:
```diff
@@ -95,32 +135,94 @@ IntersectionVertex getVertexForOsmNode(OsmNode node, OsmEntity way) {
+        var vertices = splitVerticesOnBarriers.get(node);
+        vertices.put(null, ov);
+
+        // some barrier values, such as barrier=gate, which do not block traversal will return
+        // false for node.isBarrier(), therefore we need to check the explicit absence of the
+        // barrier tag as well for the issue.
+        if (
+          !node.isBarrier() &&
+          !node.hasTag("barrier") &&
```
I guess just checking if barrier tag exists is enough.

leonardehrenfried replied:
Please add a method on `OsmNode` for this specific case and move the documentation to its Javadoc instead.

vesameskanen commented on application/src/main/java/org/opentripplanner/graph_builder/module/osm/BarrierEdgeBuilder.java:
```diff
@@ -0,0 +1,95 @@
+          var from = vs[i];
+          var to = vs[j];
+
+          if (inDegrees[i] > 0 && outDegrees[j] > 0) {
```
After the fix presented in https://github.com/opentripplanner/OpenTripPlanner/pull/6826, degree check seems unnecessary. I tested that this condition was always true when building helsinki area street graph.

miklcct replied:
I think this check is still needed to prevent unnecessary edges being built when a one-way street joins an area at a gate through a wall.

vesameskanen reviewed (commented):
I am quite happy with the functionality now, especially when https://github.com/opentripplanner/OpenTripPlanner/pull/6826 is included as well.

We are still considering if the two new issues are useful in their current form, or should the logic examine the data more carefully. For example, highways crossing a barrier tagged as `barrier=retaining wall` at a bridge structure seem to be very common. An example: https://www.openstreetmap.org/node/444209183

leonardehrenfried commented on application/src/main/java/org/opentripplanner/graph_builder/module/osm/OsmDatabase.java:
```diff
@@ -272,6 +270,12 @@ public void addWay(OsmWay way) {
 
+  private static boolean relevantForRouting(OsmWay way) {
+    return (
+      way.isRoutable() || way.isParkAndRide() || way.isBikeParking() || way.isBoardingLocation()
```
Since this method only operates on `OsmWay` that's where it should be moved.

leonardehrenfried commented on application/src/test/java/org/opentripplanner/graph_builder/module/osm/BarrierEdgeBuilderTest.java:
```diff
@@ -0,0 +1,307 @@
+import org.opentripplanner.street.model.vertex.StreetVertex;
+import org.opentripplanner.street.search.TraverseMode;
+
+class BarrierEdgeBuilderTest {
```
Very good testing!

miklcct commented:
I just found out that I made a mistake in a previous PR, which was surprising untested. The method had a test case but it didn't contain this specific case.

https://github.com/opentripplanner/OpenTripPlanner/pull/6764/files#diff-91193fccae4c9a19c4e80aafdc3c8ec8ac4de356c86e42331c6728f85459b2a6R786

I forgot to add a `return false;` statement after checking all access was denied.

I have just corrected it and added a test for it.

vesameskanen reviewed (changes requested):
I added a couple of  minor request. We are close getting this merged.
